File: problems/1022/solution.go
package main
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumRootToLeaf(root *TreeNode) int {
    if root == nil {
        return 0
    }
    
    ans, curPath := 0, 0
    
    for root != nil {
        if root.Left != nil {
            prev, steps := root.Left, 1
            
            for prev.Right != nil && prev.Right != root {
                prev = prev.Right
                steps++
            }
            
            if prev.Right == nil {
                curPath = (curPath << 1) | root.Val
                prev.Right = root
                root = root.Left
            } else {
                if (prev.Left == nil) {
                    ans += curPath
                }
                
                curPath >>= steps
                prev.Right = nil
                root = root.Right
            }
        } else {
            curPath = (curPath << 1) | root.Val
            
            if root.Right == nil {
                ans += curPath
            }
            
            root = root.Right
        }
    }
    
    return ans
}

File: problems/1022/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumRootToLeaf(self, root: TreeNode) -> int:
        if not root:
            return 0
        
        ans, q = 0, deque([(root, '')])
        
        while q:
            node, b = q.popleft()
            b += str(node.val)
            
            if not node.left and not node.right:
                ans += int(b, 2)
            
            if node.left: q.append((node.left, b))
            if node.right: q.append((node.right, b))
        
        return ans